### PR TITLE
Base uint2

### PR DESCRIPTION
**Pull request by nbougalis**
This PR builds on top of #3640 (the bottom-most commits are from there) and involves a refactoring and general code cleanup of several classes involved with SHAMaps. In addition to improving memory usage by 8 bytes per SHAMapAbstractNode, the changes also improve type safety by having distinct classes for each type of leaf we have, instead of a single class.

Each change is a separate commit (hence the volume of commits). Since this is critical code, I wanted to make sure that the changes were small enough to be easy to verify by inspection. For the same reason, I (mostly) avoided running `clang-format` and expect that there are several violations.

This is complex code and it can be challenging to grok, especially since the rationale behind some of these decisions exists only in the heads of three people: Arthur, David and myself. To that end, I made several improvements to the documentation as well, which should help make this code more tractable for others. Honestly, this is a WIP.

You only need to review the commits tagged `[WIP]` but please review them **VERY CAREFULLY**. If you're even slightly unsure that a change is correct, or if something is unclear, please leave a comment.

### CODE REVIEW

File: src/ripple/app/misc/SHAMapStoreImp.cpp [hunk 1]
```diff
@@ -302,8 +302,9 @@ SHAMapStoreImp::copyNode(
     SHAMapAbstractNode const& node)
 {
     // Copy a single record from node to dbRotating_
-    dbRotating_->fetchNodeObject(
-        node.getNodeHash().as_uint256(), node.getSeq());
+    // FIXME: node.owner() is the wrong thing to pass: fetchNodeObject expects
+    //        a ledger sequence, not a copy-on-write identifier.
+    dbRotating_->fetchNodeObject(node.getHash().as_uint256(), node.owner());
```
Comment (nbougalis) [inline src/ripple/app/misc/SHAMapStoreImp.cpp:307]: After mentioning this bug in chat, it was fixed by @miguelportilla in #3643 so it will be resolved in a different PR, and this comment will go away during the merge/conflict resolution.

**Review by scottschurr — APPROVED** [review]
I reviewed this pull request carefully, one commit at a time.  As far as I can see everything is in order.

I have two further suggestions for this pull request.

- I think now would be good time to rename `SHAMapTreeNode` to `SHAMapLeafNode`.  That seems to me to be what the class represents.

- I really like the way you renamed `SHAMap::seq_` to `cowid_`  Similarly, the rename of member variable `SHAMapAbstractNode::mSeq`  to `SHAMapAbstractNode::cowid_` is a good choice.  However I think the member function `SHAMapAbstractNode::owner()` (which returns `cowid_`) would be better named `SHAMapAbstractNode::cowid()`.  The value is easier to track and audit when it does not change names in the middle of usage.  Similarly, I think the parameters of `SHAMap*Node` member functions named `owner` would be easier to follow if they were renamed `cowid`.

Neither of these suggests affect correctness, but I think they will make the code easier to follow.

File: src/ripple/shamap/SHAMapTreeNode.h [hunk 2]
```diff
@@ -134,8 +133,14 @@ class SHAMapAbstractNode
     SHAMapAbstractNode&
     operator=(SHAMapAbstractNode const&) = delete;
 
-    SHAMapAbstractNode(SHAMapNodeType type, std::uint32_t seq);
-    SHAMapAbstractNode(SHAMapNodeType type, std::uint32_t seq, SHAMapHash const& hash);
+    explicit SHAMapAbstractNode(std::uint32_t seq) : mSeq(seq)
+    {
+    }
+
+    explicit SHAMapAbstractNode(std::uint32_t seq, SHAMapHash const& hash)
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/shamap/SHAMapTreeNode.h]: @scottschurr just to be pedantic, putting `explicit` on multi-parameter constructors prevents them from being initialized with braces. I.e. `explicit Foo(int, int)` prevents foo from being initialized with `{1,2}`. I'd probably prefer we keep the `explicit`, but I don't feel that strongly about it (and in fact I often leave it off multi-paramter constructors myself).

[the remark this answers is not in the document]
Comment (scottschurr) [reply src/ripple/shamap/SHAMapTreeNode.h]: @seelabs, thanks for that correction.  Excellent point.

File: src/ripple/app/ledger/impl/InboundLedgers.cpp [hunk 3]
```diff
@@ -166,40 +164,38 @@ class InboundLedgersImp : public InboundLedgers, public Stoppable
...
-            // useful.
-            if (packet.type() == protocol::liAS_NODE)
-            {
+            // Stash the data for later processing and see if we need to
+            // dispatch
+            if (ledger->gotData(std::weak_ptr<Peer>(peer), packet))
                 app_.getJobQueue().addJob(
-                    jtLEDGER_DATA, "gotStaleData", [this, packet_ptr](Job&) {
-                        gotStaleData(packet_ptr);
+                    jtLEDGER_DATA, "processLedgerData", [this, ledger](Job&) {
```
Comment (seelabs) [inline src/ripple/app/ledger/impl/InboundLedgers.cpp:178]: Clang warns about unnecessary capture of `this` here.

File: src/ripple/app/ledger/impl/LedgerMaster.cpp [hunk 4]
```diff
@@ -2121,56 +2166,49 @@ LedgerMaster::makeFetchPack(
...
-                        &reply,
-                        lSeq,
-                        std::placeholders::_1,
-                        std::placeholders::_2));
+            std::uint32_t lSeq = want->info().seq;
+
+            {
+                // Serialize the ledger header:
+                Serializer hdr(128);
+                hdr.erase();
```
Comment (seelabs) [inline src/ripple/app/ledger/impl/LedgerMaster.cpp:2179]: Why call `erase`?

File: src/ripple/shamap/impl/SHAMap.cpp [hunk 5]
```diff
@@ -937,34 +914,33 @@ int
...
+SHAMap::flushDirty(NodeObjectType t)
 {
-    return walkSubTree(true, t, seq);
+    // We only write back if this map is backed.
+    return walkSubTree(backed_, t);
 }
 
 int
-SHAMap::walkSubTree(bool doWrite, NodeObjectType t, std::uint32_t seq)
+SHAMap::walkSubTree(bool doWrite, NodeObjectType t)
```
Comment (seelabs) [inline src/ripple/shamap/impl/SHAMap.cpp:928]: This function has a new precondition: if `doWrite` is true, then `backed_` must be true as well. Maybe add an assert?
```
assert(!doWrite || backed_);
```

**Review by seelabs — APPROVED** [review]
:+1: Very nice set of cleanups.

File: src/ripple/app/ledger/impl/LedgerMaster.cpp (hunk 4 above)
Comment (nbougalis) [reply src/ripple/app/ledger/impl/LedgerMaster.cpp:2179]: Uhhh... stylistic flair? :sweat_smile:

File: src/ripple/app/consensus/RCLConsensus.cpp [hunk 6]
```diff
@@ -315,9 +315,8 @@ RCLConsensus::Adaptor::onClose(
         Serializer s(2048);
         tx.first->add(s);
         initialSet->addItem(
-            SHAMapItem(tx.first->getTransactionID(), std::move(s)),
-            true,
-            false);
+            SHAMapNodeType::tnTRANSACTION_NM,
```
[the remark this answers is not in the document]
Comment (nbougalis) [reply src/ripple/app/consensus/RCLConsensus.cpp:318]: :vulcan_salute:

File: src/ripple/shamap/impl/SHAMapSync.cpp [hunk 7]
```diff
@@ -632,7 +632,10 @@ SHAMap::addKnownNode(
                 return SHAMapAddNode::invalid();
             }
 
-            if (!newNode->isInBounds(iNodeID))
+            // Inner nodes must be at a level strictly less than 64
+            // but leaf nodes (while notionally at level 64) can be
+            // at any depth:
+            if (newNode->isInner() && iNodeID.getDepth() == 64)
```
[the remark this answers is not in the document]
Comment (nbougalis) [reply src/ripple/shamap/impl/SHAMapSync.cpp:624]: Good idea.

File: src/ripple/app/ledger/impl/LedgerMaster.cpp [hunk 8]
```diff
@@ -2033,6 +2033,51 @@ LedgerMaster::gotFetchPack(bool progress, std::uint32_t seq)
...
+        have,
+        [withLeaves, &max, into, seq](SHAMapAbstractNode const& smn) -> bool {
+            if (withLeaves || smn.isInner())
+            {
+                Serializer s;
+                smn.serializeWithPrefix(s);
+
+                protocol::TMIndexedObject* obj = into->add_objects();
+                obj->set_ledgerseq(seq);
+                obj->set_hash(smn.getHash().as_uint256().data(), 256 / 8);
```
[the remark this answers is not in the document]
Comment (nbougalis) [reply src/ripple/app/ledger/impl/LedgerMaster.cpp]: I'll do you one better: `smn.getHash().as_uint256().size()` :wink:

File: src/ripple/app/ledger/impl/LedgerMaster.cpp (hunk 4 above)
Comment (nbougalis) [reply src/ripple/app/ledger/impl/LedgerMaster.cpp:2182]: Moved `hdr` out of the loop.

File: src/ripple/basics/base_uint.h [hunk 9]
```diff
@@ -363,99 +373,62 @@ class base_uint
...
-        unsigned char* pOut = end() - ((pEnd - pBegin + 1) / 2);
+        auto parseHexChar = [](char c) {
+            if (auto x = charUnHex(c); x != -1)
+                return static_cast<unsigned char>(x);
 
-        *this = beast::zero;
-
-        if ((pEnd - pBegin) & 1)
-            *pOut++ = charUnHex(*pBegin++);
+            throw c;
```
Comment (HowardHinnant) [inline src/ripple/basics/base_uint.h:400]: I like this improved API for `base_uint`, but I don't recommend using exceptions for flow control.  It is expensive both in terms of performance and code size.  Furthermore the `static_cast` is in the wrong place to be effective since the expression `(hi << 4) + lo` promotes everything back to `int`.

Here is what I recommend:

```c++
    [[nodiscard]] bool
    parseHex(std::string_view sv)
    {
        zero();

        if (sv == "0")
            return true;

        if (sv.size() != bytes * 2)
            return false;

        auto out = data();

        auto in = sv.begin();

        while (in != sv.end())
        {
            auto const hi = charUnHex(*in++);
            auto const lo = charUnHex(*in++);
            if (hi == -1 || lo == -1)
                return false;
            *out++ = static_cast<value_type>((hi << 4) + lo);
        }

        return true;
    }
```

**Review by scottschurr — APPROVED** [review]
Looks great!  I left a few notes on the changes since my last review.  Also, I think `Message::getBufferSize()` could be `const`.  But none of the notes I left are mandatory.  Thanks for your work here!

**Comment by seelabs** [comment]
The CI issues need to be resolved as well (mostly looks like missing `override`s).

File: src/ripple/shamap/SHAMapTreeNode.h [hunk 10]
```diff
@@ -301,35 +301,21 @@ class SHAMapInnerNode : public SHAMapAbstractNode,
...
-    clone(std::uint32_t seq) const override;
-
-    SHAMapNodeType
-    getType() const override
-    {
-        return mType;
-    }
 
     bool
     isLeaf() const override
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/shamap/SHAMapTreeNode.h]: Reopening this so this comment is seen. Marking these function `final` without the `override` is causing gcc8 to warn; we need both the `final` and the `override`. However, some of these classes are already marked as `final`. For `final` classes, I would not mark these functions as `final` (just to cut down on the noise of `final override`). I'll make a patch to resolve the warnings.

[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/shamap/SHAMapTreeNode.h]: @nbougalis 
Here's a patch to reformat with clang-format: https://github.com/seelabs/rippled/commit/e9c5db58921b5292bd3571fd9ee8965c53dc7e3f
Here's a patch to resolve the `suggest-override` warnings: https://github.com/seelabs/rippled/commit/814a52959fc8f409d6578c1def827884487807c9

Hopefully this will resolve the CI issues. Also note, I'm not able to test those locally, as I do not have access to gcc 8.

**Review by HowardHinnant — APPROVED** [review]

File: src/ripple/shamap/README.md [hunk 11]
```diff
@@ -309,9 +309,9 @@ the following data:
...
 4.  Flag to aid online delete and consistency with data on disk.
 
-## SHAMapTreeNode ##
+## SHAMapLeafNode ##
 
-SHAMapTreeNode publicly inherits directly from SHAMapAbstractNode.  It holds the
+SHAMapLeafNode publicly inherits directly from SHAMapAbstractNode.  It holds the
 following data:
 
 1.  A shared_ptr to a const SHAMapItem.
```
[the remark this answers is not in the document]
Comment (nbougalis) [reply src/ripple/shamap/README.md:317]: Will look.

File: src/ripple/shamap/SHAMapTreeNode.h [hunk 12]
```diff
@@ -196,7 +196,7 @@ class SHAMapAbstractNode
 
     /** Make a copy of this node, setting the owner. */
```
[the remark this answers is not in the document]
Comment (nbougalis) [reply src/ripple/shamap/SHAMapTreeNode.h:197]: Yes.

**Comment by seelabs** [comment]
We've got a linker error on mac: 
```
Undefined symbols for architecture x86_64:

  "std::__1::__itoa::__u64toa(unsigned long long, char*)", referenced from:

      std::__1::__itoa::__traits_base<unsigned long long, void>::__convert(unsigned long long, char*) in libxrpl_core.a(unity_6_cxx.cxx.o)
```

**Comment by HowardHinnant** [comment]
This comes from a new use of `std::to_chars` and does not happen except on platforms where the macOS is older than  current-3 (I'm testing on current-3 and am linking).  Apple simply does not ship `std::to_chars` on the older versions of macOS.

I think this is safe to ignore.

**Comment by scottschurr** [comment]
Great continues to look even better.  I found one last remaining reference to `SHAMapAbstractNode` in a comment in SHAMap.cpp.  It looks like this:
```
    @note The node must have already been unshared by having the caller
          first call SHAMapAbstractNode::share().
```
I think if you change that to `SHAMapTreeNode::share()` you'll have it all cleaned out.

**Comment by seelabs** [comment]
@nbougalis This patch should fix the CI linker error: https://github.com/ripple/rippled/pull/3661/commits/eae43672281fa3601dae35ddb10a7c32b5e1e55e

**Comment by nbougalis** [comment]
I've addressed all remaining concerns.

**Review by seelabs — COMMENTED** [review]
@nbougalis The mac linker errors still appear to be there. We need the patch referenced in this comment: https://github.com/ripple/rippled/pull/3646#issuecomment-730427177

**Review by gregtatcam — APPROVED** [review]
LGTM :+1:
